Initial attempt at making a more complete loop guide (#7)
DenizenScript/Denizen-Beginners-Guide: source/guides/basics/loops.md

Do That Again: Loops (PARTIAL)
--------------------

> **TODO: Write-up that introduces loops in general, using the repeat command.**

Reviewer: Missing table of contents (look at source of other pages)

All the scripts up until now had a single directional flow from top to bottom. An exact order was listed and were executed one after the other, with maybe skipping a few if you had an `if` command. However, what if you wanted to do that script you just wrote again? Well you have a few options. The most straightforward approach would simply be writing that same bit of script again. While this may work if you only wanted it repeated twice, or if there is only one line you want repeated, this can get out of hand really quickly.

Luckily you don't have to do this, Denizen has 3 different loop commands that repeat a block of script:
- `repeat` to repeat a block a set amount of times
- `foreach` to repeat a block based on a list input
- `while` keep repeating a block until a condition is met

### The Repeat Command

The most basic use of a loop is to simply just do a block of script again, in which the repeat command is best suited for.
The basic format of a repeat command is

> ```

Reviewer: `dscript_blue`

- repeat (number of times):
    - (commands here)
    - (will loop number of times)
```

> This structure is similar to the `if` command, where you end the `repeat` command with a `:`. The commands that get looped need to also be indented in to let denizen know that these are the commands you want to be looped.

Reviewer: `Denizen`, capital D

Here is what it would look like in a real script...
```dscript_green
my_zap_task:
    type: task
    script:
    - repeat 5:
        - strike <player.location>
        - narrate "you have been zapped!"
        - wait 1s

>     - narrate "no more zaps"

Reviewer: `zap` is a command in Denizen, a different word should be used here to avoid risk of confusion - probably `lightning` / `you have been struck by lightning`

```

> This would in essence loop 5 times, where the player would be stricken by lightning 5 times as well as been told "you have been zapped!" 5 times.

Reviewer: `struck` not `stricken`
`be` not `been`

Reviewer: Actually this should be future tense not past - `will be` rather than `would be` - same for most other phrasing through this page

> There is also a `wait` command at the end of the block. This pauses the script for a certain amount of time (in this example 1 second). This will be discussed more in a later guide. For a now its here so that you are not zapped 5 times at the same time.

Reviewer: `wait` commands are explained earlier in the guide than this page, so this addendum is unnecessary.

Do That Again: Loops
--------------------

```eval_rst
.. contents:: Table of Contents
    :local:
```

### What Is A 'Loop'?

All the scripts up until now had a single directional flow from top to bottom. An exact order was listed and were executed one after the other, with maybe skipping a few if you had an `if` command. However, what if you wanted to do that script you just wrote again? Well you have a few options. The most straightforward approach would simply be writing that same bit of script again. While this may work if you only wanted it repeated twice, or if there is only one line you want repeated, this can get out of hand really quickly.

Luckily you don't have to do this, Denizen has 3 different loop commands that repeat a block of script:
- `repeat` to repeat a block a set amount of times
- `foreach` to repeat a block based on a list input
- `while` keep repeating a block until a condition is met

### The Repeat Command

> The most basic use of a loop is to simply just do a block of script again, in which the repeat command is best suited for.

Reviewer 2: ```suggestion
The most basic use of a loop is to simply run a block of commands again, which the repeat command is best suited for.
```

The basic format of a repeat command is
```dscript_blue
- repeat (number of times):
    - (commands here)
    - (will loop number of times)
```

This structure is similar to the `if` command, where you end the `repeat` command with a `:`. The commands that get looped need to also be indented in to let Denizen know that these are the commands you want to be looped.

Here is what it would look like in a real script...
```dscript_green
my_lightning_task:
    type: task
    script:
    - repeat 5:
        - strike <player.location>
        - narrate "you have been struck by lightning!"
        - wait 1s
    - narrate "no more lightning"
```

This would in essence loop 5 times, where the player would be struck by lightning 5 times as well as be told "you have been struck by lightning!" 5 times.
There is also a `wait` command at the end of the block. This pauses the script for a certain amount of time (in this example 1 second). This is important in a lot of cases so that it does not happen all at once. Without the `wait`, all 5 loops will run at the same time.

Once the loop is done, it then proceeds to continue with the rest of the script - ending with "no more lightning"

> Now what if you wanted to let the player know how many times they have been struck? Granted you could put in a definition that counts up each loop, but `repeat` already does that for you. All you need to add is the `as:` argument to the repeat command like so.

Reviewer 2: ```suggestion
This would loop 5 times, wherein the player would be struck by lightning as well as be told "you have been struck by lightning!"
There is also a `wait` command at the end of the block. This pauses the script for a certain amount of time (in this example 1 second). This is important in many cases so that the commands don't run instantly after one another. In this example, it would seem as if the lightning strikes happened all at once.

Once the loop is done, the script continues, ending with "no more lightning."

What if you wanted to let the player know how many times they have been struck? You could add a definition that tracks the amount of loops, but `repeat` already does that for you. All you need to add is the `as:` argument to the repeat command like so:
```

```dscript_green
my_lightning_task:
    type: task
    script:
    - repeat 5 as:count:
        - strike <player.location>
        - narrate "you have been struck by lightning <[count]> times!"
        - wait 1s
    - narrate "no more lightning"
```

> The narrate would now say "you have been struck by lightning 1 times!", "you have been struck by lightning 2 times!", etc... up to 5.

Reviewer 2: ```suggestion
This would now narrate "you have been struck by lightning (n) times!", where `n` is the amount of loops; 1 to 5.
```

Author: This doesn't feel right as a description of whats going on... this is just replacing <[count]> with (n)

You can try these examples in-game via `/ex run my_lightning_task`.

### The Foreach Command

> While the repeat command is handy, a lot of the times you would find yourself wanting to loop over the contents of a list, in that for each item in the list, do something.

Reviewer 2: ```suggestion
While the repeat command is handy, you'll often find yourself wanting to loop over the contents of a list and do something with each item in it.
```

Let's say we want to tell the player where all the cows are around them

> Now, you can use the repeat command for this by doing this...

Reviewer 2: ```suggestion
Let's say we want to tell the player where all the cows around them are located.
With the repeat command, you might do something like this:
```

```dscript_red
my_cow_task:
    type: task
    script:
    - define cows <player.location.find_entities[cow].within[30]>
    - repeat <[cows].size> as:index:
        - define cow <[cows].get[<[index]>]>
        - narrate "There's a cow just <[cow].location.distance[<player.location>].round> blocks away!"
        - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3
```

However, there is a much better way using the `foreach` command.

> `foreach` has the same structure as repeat
> ```dscript_blue
> - foreach (some list) as:(definition to store item):
>     - (commands here)
>     - (will loop per item in list)

Reviewer 2: ```suggestion
```dscript_blue
# foreach has the same structure as repeat
- foreach (some list) as:(definition to store item):
    - (commands here)
    - (will loop per item in list)
```

Author: why would you put this inside the code block?

Reviewer 2: idk

Reviewer 2: Just attach it to the main paragraph instead

```

> Now lets see how we can simplify that script with a `foreach` loop

Reviewer 2: ```suggestion
Here's how we can simplify that script with a `foreach` loop:
```

In addition, foreach has a built-in definition of `<[loop_index]>` that keeps track of how many times it has looped.

For instance...
```dscript_green
my_cow_task:
    type: task
    script:
    - foreach <player.location.find_entities[cow].within[30]> as:cow:
        - narrate "<[loop_index]> - There's a cow just <[cow].location.distance[<player.location>].round> blocks away!"
        - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3
```

The narrate would now be prefaced with a "1 - ", "2 - ", etc... as the loop goes on.

You can try these examples in-game via `/ex run my_cow_task` while standing near some cows.

### The While Command

There may be a few times where neither the repeat nor the foreach command work, because you don't know how many times it should loop, you just have a condition of when to stop.

This can be thought of as an `if` command but as a loop. The difference is that the `if` command simply runs the block if its true, the `while` command run the block as long as the condition is true 
```
- while (condition):
    - (commands here)
    - (will loop while condition is true)
```

> The condition format is the same format as the `if` command, see that guide page for more details on the structure.

Reviewer: should link the other guide page when it's referenced

        - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3
```

Much better. Most notably, we don't have have to get the item `cow` from the list, its already defined.

> In addition, foreach has a built-in definition of `<[loop_index]>` that keeps track of how many times it has looped.

Reviewer 2: ```suggestion
Much better. Most notably, we don't have have to get the item `cow` from the list: it's already defined.
Additionally, foreach has a built-in definition of `<[loop_index]>` that keeps track of how many times it has looped.
```

For instance...
```dscript_green
my_cow_task:
    type: task
    script:
    - foreach <player.location.find_entities[cow].within[30]> as:cow:
        - narrate "<[loop_index]> - There's a cow just <[cow].location.distance[<player.location>].round> blocks away!"
        - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3
```

> The narrate would now be prefaced with a "1 - ", "2 - ", etc... as the loop goes on.

Reviewer 2: This seems unnecessary

Author: I figured it would be useful to demonstrate, as a counterpoint why someone *might* want to use a repeat loop over foreach when looping a list

You can try these examples in-game via `/ex run my_cow_task` while standing near some cows.

### The While Command

There may be a few times where neither the repeat nor the foreach command work, because you don't know how many times it should loop, you just have a condition of when to stop.

> This can be thought of as an `if` command but as a loop. The difference is that the `if` command simply runs the block if its true, the `while` command run the block as long as the condition is true 

Reviewer 2: ```suggestion
You might encounter a situation where you don't know how many times you're going to loop, but you know when to stop. You can achieve this with a while loop. It's like an `if` command that runs the block below it as long as the condition is true.
```

```dscript_blue
- while (condition):
    - (commands here)
    - (will loop while condition is true)
```

> The condition format is the same format as the `if` command, see that guide page for more details on the structure. [if command](/guides/basics/if-command)

Reviewer 2: ```suggestion
The condition format is the same as the `if` command; [see that guide page](/guides/basics/if-command) for more details on its structure.
```

        - wait 2s
```

You can try this in-game via `/ex run my_move_task`. This task when run will keep telling you to move away until you have moved 3 blocks away from when you first were.

Just like the `repeat` command we have a `wait` command at the end of the loop. While it can be optional for `repeat` loops to use this, it is almost always required for `while` loops. Without it it will try to check as much as it can all at the same time, which can cause the server to crash.

> As a word of warning, `while` loops should be avoid if possible. It is very easy to make what is known as an infinite loop, in that the loop will never have a chance to stop. Once its running, it is very hard to force it to stop beyond just stopping the server.

Reviewer: `avoid` should be `avoided`
`in that` should be `as in`

        - wait 2s
```

> You can try this in-game via `/ex run my_move_task`. This task when run will keep telling you to move away until you have moved 3 blocks away from when you first were.

Reviewer 2: ```suggestion
You can try this in-game via `/ex run my_move_task`. This task, when run, will repeatedly tell you to move away until you have moved 3 blocks from where you first were.
```

> Just like the `repeat` command we have a `wait` command at the end of the loop. While it can be optional for `repeat` loops to use this, it is almost always required for `while` loops. Without it it will try to check as much as it can all at the same time, which can cause the server to crash.
> 
> As a word of warning, `while` loops should be avoided if possible. It is very easy to make what is known as an infinite loop, as in the loop will never have a chance to stop. Once its running, it is very hard to force it to stop beyond just stopping the server.

Reviewer 2: ```suggestion
Just like the `repeat` command, we have a `wait` command at the end of the loop. While it's only useful in some cases with `repeat` loops, you should almost always use it with `while` loops. Without one, it will try to check the condition as often as possible all at the same time, which can cause the server to crash.

A word of warning: `while` loops should be avoided if possible. It is very easy to make what is known as an "infinite loop", as in a loop that never has a chance to stop. Once it's running, it's difficult to force it to stop beyond just stopping the server.
```

### Stop The Loop

> Sometimes in loops you only want to keep looping until you reach a certain point, however not necessarily stop the script. While `stop` is used to stop scripts entirely, `repeat/foreach/while stop` is used to stop the relevant loop

Reviewer 2: ```suggestion
Sometimes in loops you only want to keep looping until you reach a certain point. However, you may not want to stop the script entirely. While `stop` is used to stop scripts, `repeat/foreach/while stop` is used to stop the relevant loop.
```

> For instance in the first example we had, if we wanted to stop sending lightning if the player had less than 5 health we would do as follows

Reviewer 2: ```suggestion
For instance, in the first example, if we wanted to stop striking lightning if the player had less than 5 health, we would do as follows:
```

```dscript_green
my_lightning_task:
    type: task
    script:
    - repeat 5 as:count:
        - strike <player.location>
        - narrate "you have been struck by lightning <[count]> times!"
        - if <player.health> < 5:
            - repeat stop
        - wait 1s
    - narrate "no more lightning"
```

> Take note, only the loop stops, the end `narrate` line still runs.

Reviewer 2: ```suggestion
Since we're using `repeat stop`, only the loop ends; the end `narrate` line still runs.
```

It can also be used in `foreach` loops. For instance, if we wanted to stop looping if we found a baby cow in the example above

```dscript_green
my_cow_task:
    type: task
    script:
    - foreach <player.location.find_entities[cow].within[30]> as:cow:
        - narrate "<[loop_index]> - There's a cow just <[cow].location.distance[<player.location>].round> blocks away!"
        - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3
        - if <[cow].is_baby>:
            - narrate "omg its a baby!!"
            - foreach stop

> ```

Reviewer 2: Unnecessary as it is said that it works for foreach and while loops

Author: I thought it would be good to demonstrate in multiple examples, at the very least to make it clear of the syntax

### Next Please

> There are times where you want to skip to the next loop without running the rest of the current loop. For that you use `repeat/foreach/while next`

Reviewer 2: ```suggestion
There are times when you want to skip to the next loop without finishing the current iteration. For this, you would use `repeat/foreach/while next`.
```

Take the `foreach` example we had, lets say we want to skip over all baby cows. While you can put the whole block inside an `if` command like so...
```dscript_green
my_cow_task:
    type: task
    script:
    - foreach <player.location.find_entities[cow].within[30]> as:cow:
        - if !<[cow].is_baby>:
            - narrate "<[loop_index]> - There's a cow just <[cow].location.distance[<player.location>].round> blocks away!"
            - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3

> ```

Reviewer 2: Unnecessary - get to the point

You could instead use `next`

```dscript_green
my_cow_task:
    type: task
    script:
    - foreach <player.location.find_entities[cow].within[30]> as:cow:
        - if <[cow].is_baby>:
            - foreach next
        - narrate "<[loop_index]> - There's a cow just <[cow].location.distance[<player.location>].round> blocks away!"
        - playeffect effect:fireworks_spark at:<[cow].location> visibility:50 quantity:100 data:0 offset:3
```

> 

Reviewer 2: This isn't a great example, as an if statement probably works better here.

Author: does it need to be a great example? Just needs to get the point across...

Reviewer 2: This segment is a demonstration of the usefulness of the next argument, so yes the examples should make sense in my opinion
